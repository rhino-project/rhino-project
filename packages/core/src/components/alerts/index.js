import React from 'react';
import PropTypes from 'prop-types';

import { Alert } from 'reactstrap';

export const DetailAlert = ({
  title,
  description,
  color = 'primary',
  children
}) => {
  return (
    <Alert color={color}>
      <h6>{title}</h6>
      {description && <p>{description}</p>}
      {children}
    </Alert>
  );
};

DetailAlert.propTypes = {
  children: PropTypes.node,
  color: PropTypes.string,
  description: PropTypes.node,
  title: PropTypes.string.isRequired
};

export const SuccessAlert = (props) => (
  <DetailAlert {...props} color="success" />
);

export const DangerAlert = (props) => <DetailAlert {...props} color="danger" />;
