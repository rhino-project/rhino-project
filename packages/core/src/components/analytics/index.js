export * from './IdentityAnalytics';
export * from './PageAnalytics';
