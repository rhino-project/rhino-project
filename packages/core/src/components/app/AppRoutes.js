import React from 'react';
import { Route, Switch, useRouteMatch } from 'react-router-dom';

import { customRoutes } from 'routes/custom';
import NotFoundPage from 'rhino/components/shared/NotFoundPage';
import { ENABLE_MODEL_ROUTES } from 'config';
import { accountSettingsRoute, settingsRoute } from '../../routes';
import { modelRoutes } from '../../routes/model';

export const AppRoutes = () => {
  const match = useRouteMatch();

  return (
    <Switch>
      {accountSettingsRoute(match)}
      {settingsRoute(match)}
      {customRoutes(match)}
      {ENABLE_MODEL_ROUTES ? modelRoutes(match) : []}
      <Route component={NotFoundPage} />
    </Switch>
  );
};
