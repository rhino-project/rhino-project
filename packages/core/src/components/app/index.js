export * from './AccountMenu';
export * from './BaseOwnerSwitcher';
export * from './NotificationMenu';
export * from './Toaster';
