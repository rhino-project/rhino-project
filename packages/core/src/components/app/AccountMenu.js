import { useMemo } from 'react';
import { NavLink } from 'react-router-dom';
import {
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown
} from 'reactstrap';

import { NavIcon } from 'rhino/components/icons';
import { useUser } from 'rhino/hooks/auth';
import { useBaseOwner, useHasRoleOf } from 'rhino/hooks/owner';
import { useAccountSettingsPath, useSettingsPath } from 'rhino/hooks/routes';
import { useSignOutAction } from 'rhino/queries/auth';
import { hasOrganizationsModule } from 'rhino/utils/models';

const OrganizationSettings = () => {
  const baseOwner = useBaseOwner();
  const settingsPath = useSettingsPath();

  return (
    <>
      <DropdownItem divider />
      <DropdownItem tag={NavLink} to={`${settingsPath}/profile`}>
        {baseOwner?.name} Settings
      </DropdownItem>
    </>
  );
};

const AccountMenu = () => {
  const { mutate: signOutAction } = useSignOutAction();
  const user = useUser();
  const accountSettingsPath = useAccountSettingsPath();
  const isAdmin = useHasRoleOf('admin');
  const showOrgSettings = useMemo(
    () => hasOrganizationsModule() && isAdmin,
    [isAdmin]
  );

  return (
    <UncontrolledDropdown nav direction="up">
      <DropdownToggle
        id="account-menu"
        className="d-flex align-items-center text-light no-arrow"
        nav
        caret
      >
        <NavIcon icon="person-circle" extraClass="flex-shrink-0" />
        <span className="d-block mx-2 text-truncate flex-grow-1">
          {user?.nickname || user?.name?.split(' ')[0]}
        </span>
        <NavIcon icon="chevron-down" extraClass="flex-shrink-0" />
      </DropdownToggle>
      <DropdownMenu dark end>
        <DropdownItem
          id="account-settings"
          tag={NavLink}
          to={`${accountSettingsPath}/profile`}
        >
          Account Settings
        </DropdownItem>
        {showOrgSettings && <OrganizationSettings />}
        <DropdownItem divider />
        <DropdownItem id="account-signout" onClick={signOutAction}>
          Sign Out
        </DropdownItem>
      </DropdownMenu>
    </UncontrolledDropdown>
  );
};

export default AccountMenu;
