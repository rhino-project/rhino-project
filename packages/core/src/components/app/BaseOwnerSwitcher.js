import {
  DropdownItem,
  DropdownMenu,
  DropdownToggle,
  UncontrolledDropdown
} from 'reactstrap';
import { NavIcon } from 'rhino/components/icons';
import { useBaseOwnerNavigation } from 'rhino/hooks/history';
import { useBaseOwner, useBaseOwnerId, useUserRoles } from 'rhino/hooks/owner';
import { useRootPath } from 'rhino/hooks/routes';

const BaseOwnerSwitcher = () => {
  const baseOwnerId = useBaseOwnerId();
  const baseOwnerNavigation = useBaseOwnerNavigation();
  const usersRoles = useUserRoles();
  const baseOwner = useBaseOwner();
  const rootPath = useRootPath();

  // Only show the dropdown if there is more than one possible base owner
  if (!baseOwnerId || !usersRoles || usersRoles.length <= 1) {
    return null;
  }

  const handleClick = (baseOwnerClicked) =>
    baseOwnerNavigation.push(rootPath, baseOwnerClicked.id);

  return (
    <UncontrolledDropdown nav direction="up">
      <DropdownToggle
        nav
        caret
        className="d-flex align-items-center text-light no-arrow"
      >
        <NavIcon icon="building" extraClass="flex-shrink-0" />
        <span className="d-block ms-2 overflow-hidden flex-grow-1">
          {baseOwner?.name}
        </span>
        <NavIcon icon="chevron-down" extraClass="flex-shrink-0" />
      </DropdownToggle>
      <DropdownMenu dark end>
        {usersRoles.map((ur) => (
          <DropdownItem
            key={ur.organization.id}
            onClick={() => handleClick(ur.organization)}
          >
            {ur.organization.name}
          </DropdownItem>
        ))}
      </DropdownMenu>
    </UncontrolledDropdown>
  );
};

export default BaseOwnerSwitcher;
