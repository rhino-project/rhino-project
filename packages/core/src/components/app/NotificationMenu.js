import React from 'react';
import {
  Badge,
  UncontrolledDropdown,
  DropdownItem,
  DropdownMenu,
  DropdownToggle
} from 'reactstrap';
import PropTypes from 'prop-types';
import { NavLink } from 'react-router-dom';
import classnames from 'classnames';

import {
  useNotifications,
  useNotificationsOpenAll,
  useNotificationsOpen
} from 'rhino/queries/notifications';
import { NavIcon } from 'rhino/components/icons';

const NotificationMenu = ({ sidebarMode = false }) => {
  const { data: { data: notifications } = {}, refetch } = useNotifications();
  const { mutate: openAll } = useNotificationsOpenAll();
  const { mutate: openOne } = useNotificationsOpen();
  const hasNotifications = notifications?.count > 0;

  const handleItemClick = (notificationId) =>
    openOne(notificationId, {
      onSuccess: () => refetch()
    });
  const handleClick = () => openAll({ onSuccess: () => refetch() });

  return (
    <UncontrolledDropdown nav inNavbar direction={sidebarMode ? 'up' : 'down'}>
      <DropdownToggle nav caret className="d-flex align-items-center">
        <NavIcon icon="bell" extraClass="flex-shrink-0" />
        <span
          className={classnames('d-block', 'overflow-hidden', 'flex-grow-1', {
            'd-md-none': !sidebarMode
          })}
        >
          Notifications
        </span>

        {hasNotifications && <Badge>{notifications?.count}</Badge>}
      </DropdownToggle>
      <DropdownMenu end={sidebarMode ? false : true}>
        {notifications?.notifications?.map((n) => {
          return (
            <DropdownItem
              key={n.id}
              tag={NavLink}
              to={n.notifiable_path}
              onClick={() => handleItemClick(n.id)}
            >
              {n.printable_notifiable_name}
            </DropdownItem>
          );
        })}
        {hasNotifications ? (
          <>
            <DropdownItem divider />
            <DropdownItem disabled={!hasNotifications} onClick={handleClick}>
              Mark All Opened
            </DropdownItem>
          </>
        ) : (
          <DropdownItem disabled>
            <em>No unread notifications</em>
          </DropdownItem>
        )}
      </DropdownMenu>
    </UncontrolledDropdown>
  );
};

export default NotificationMenu;

NotificationMenu.propTypes = {
  sidebarMode: PropTypes.bool
};
