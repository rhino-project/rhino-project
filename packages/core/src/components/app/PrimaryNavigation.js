import CustomPrimaryNavigation from 'components/app/CustomPrimaryNavigation';
import { map } from 'lodash';
import PropTypes from 'prop-types';
import { useMemo } from 'react';
import { NavItem, NavSection } from 'rhino/components/nav';
import { useRhinoConfig } from 'rhino/config';
import { getBaseOwnedModels } from 'rhino/utils/models';
import { getModelIndexPath } from 'rhino/utils/routes';

const modelsRoute = (model) => getModelIndexPath(model);

const useNavModels = () => {
  const models = getBaseOwnedModels();
  return useMemo(() => models, [models]);
};

const PrimaryNavigation = ({ title = 'Resources', className, itemClass }) => {
  const { enableModelRoutes } = useRhinoConfig();
  const models = useNavModels().filter(
    (e) => e.model !== 'users_role' && e.model !== 'users_role_invite'
  );

  return (
    <>
      <CustomPrimaryNavigation className={className} itemClass={itemClass} />
      {enableModelRoutes && (
        <NavSection title={title} className={className}>
          {map(models, (m) => (
            <NavItem
              key={m.model}
              title={m.pluralReadableName}
              to={modelsRoute(m)}
              icon="list"
              className="px-3"
              extraClass={itemClass}
            />
          ))}
        </NavSection>
      )}
    </>
  );
};

export default PrimaryNavigation;

PrimaryNavigation.propTypes = {
  title: PropTypes.string,
  className: PropTypes.string,
  itemClass: PropTypes.string
};
