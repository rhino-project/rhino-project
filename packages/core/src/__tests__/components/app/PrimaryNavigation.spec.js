import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { render } from '@testing-library/react';
import { PrimaryNavigation } from '../../../components/app';
import { createWrapper } from '../../shared/helpers';
import rhinoConfig from 'rhino.config';

vi.mock('../../../hooks', async (importOriginal) => {
  const actual = await importOriginal();
  return {
    ...actual,
    useRoles: vi.fn().mockReturnValue(['admin'])
  };
});

describe('PrimaryNavigation', () => {
  const Wrapper = ({ children, ...props }) => {
    const queryClient = new QueryClient();

    return (
      <MemoryRouter {...props}>
        <QueryClientProvider client={queryClient}>
          <Routes>
            <Route path="/:baseOwnerId/*" element={<>{children}</>} />
          </Routes>
        </QueryClientProvider>
      </MemoryRouter>
    );
  };

  it('renders without crashing', () => {
    const { asFragment } = render(<PrimaryNavigation />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('disables model routes', () => {
    const spy = vi
      .spyOn(rhinoConfig, 'enableModelRoutes', 'get')
      .mockImplementation(() => false);

    const { asFragment } = render(<PrimaryNavigation />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();

    spy.mockRestore();
  });

  it('displays models passed as string of arrays', () => {
    const { asFragment } = render(<PrimaryNavigation models={['user']} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models returned from function as string of arrays', () => {
    const models = () => ['user'];
    const { asFragment } = render(<PrimaryNavigation models={models} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models passed as an object', () => {
    const { asFragment } = render(
      <PrimaryNavigation models={{ admin: ['blog'] }} />,
      {
        wrapper: createWrapper(Wrapper, {
          initialEntries: ['/1']
        })
      }
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('displays models returned from function as an object', () => {
    const models = () => {
      ['user'];
    };

    const { asFragment } = render(<PrimaryNavigation models={models} />, {
      wrapper: createWrapper(Wrapper, {
        initialEntries: ['/1']
      })
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
