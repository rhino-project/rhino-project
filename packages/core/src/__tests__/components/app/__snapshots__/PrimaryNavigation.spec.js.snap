// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PrimaryNavigation > disables model routes 1`] = `<DocumentFragment />`;

exports[`PrimaryNavigation > displays models passed as an object 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/blogs"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Blogs
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PrimaryNavigation > displays models passed as string of arrays 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/users"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Users
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PrimaryNavigation > displays models returned from function as an object 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    />
  </div>
</DocumentFragment>
`;

exports[`PrimaryNavigation > displays models returned from function as string of arrays 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/users"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Users
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PrimaryNavigation > renders without crashing 1`] = `
<DocumentFragment>
  <div>
    <div
      class="nav-sidebar-section d-flex justify-content-between align-items-center px-0 mt-3 mb-1 text-capitalize text-gray-400"
    >
      <span
        class="fw-medium"
      >
        Resources
      </span>
    </div>
    <ul
      class="nav flex-column"
    >
      <li
        class="nav-item"
      >
        <a
          class="px-3 nav-link"
          href="/1/blogs"
        >
          <div
            class="d-flex align-items-center"
          >
            <svg
              class="nav-icon me-1"
              fill="currentColor"
            >
              <use
                xlink:href="/node_modules/bootstrap-icons/bootstrap-icons.svg#list"
              />
            </svg>
            <div
              class="ms-2"
            >
              Blogs
            </div>
          </div>
        </a>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;

exports[`PrimaryNavigation > should render with global override shorthand 1`] = `
<DocumentFragment>
  <div>
    Bar
  </div>
</DocumentFragment>
`;
